ext/MitsubaLoader/CMitsubaMaterialCompilerFrontend.cpp

```diff
@@ -0,0 +1,597 @@
+            getSpectrumOrTexture(current->coating.sigmaA, node->sigmaA);
+            node->eta = IR::INode::color_t(eta);
+            node->shadowing = IR::CCoatingBSDFNode::EST_SMITH;
+            if (currType == CElementBSDF::ROUGHCOATING)
+            {
+                node->ndf = ndfMap[current->coating.distribution];
+                getFloatOrTexture(current->coating.alphaU, node->alpha_u);
+                if (node->ndf == IR::CMicrofacetSpecularBSDFNode::ENDF_ASHIKHMIN_SHIRLEY)
+                    getFloatOrTexture(current->coating.alphaV, node->alpha_v);
+                else
+                    node->alpha_v = node->alpha_u;
+            }
+            else
+            {
+                node->setSmooth();
+            }
+            bsdfQ.push(current->coating.bsdf[0]);
+        }
+        break;
+        case CElementBSDF::BLEND_BSDF:
+        {
+            nextSym = ir->allocNode<IR::CBSDFBlendNode>();
+            nextSym->children.count = 2u;
+
+            auto* node = static_cast<IR::CBSDFBlendNode*>(nextSym);
+            getFloatOrTexture(current->blendbsdf.weight, node->weight);
+
+            bsdfQ.push(current->blendbsdf.bsdf[1]);
+            bsdfQ.push(current->blendbsdf.bsdf[0]);
+        }
+        break;
+        case CElementBSDF::MIXTURE_BSDF:
+        {
+            nextSym = ir->allocNode<IR::CBSDFMixNode>();
+            auto* node = static_cast<IR::CBSDFMixNode*>(nextSym);
+            const size_t cnt = current->mixturebsdf.childCount;
+            nextSym->children.count = cnt;
+            for (int32_t i = cnt-1u; i >= 0; --i)
+                node->weights[i] = current->mixturebsdf.weights[i];
+
+            for (int32_t i = cnt-1u; i >= 0; --i)
+                bsdfQ.push(current->mixturebsdf.bsdf[i]);
+        }
+        break;
+        }
+
+        IR::INode* newParent = nullptr;
+        if (nextSym->children)
+            nodeQ.push(nextSym);
+        if (!--childrenCountdown && !nodeQ.empty())
+        {
+            newParent = nodeQ.front();
+            nodeQ.pop();
+        }
+
+        parent->children[childrenCountdown] = nextSym;
+        if (newParent)
+        {
+            parent = newParent;
+            childrenCountdown = parent->children.count;
+        }
+    }
+
+    static_cast<IR::CMaterialNode*>(root)->opacity = opacity;
+    static_cast<IR::CMaterialNode*>(root)->thin = thin;
+
+    IR::INode* surfParent = root;
+    if (surfParent->children[0]->symbol == IR::INode::ES_GEOM_MODIFIER)
+        surfParent = surfParent->children[0];
+
+    IR::INode::children_array_t surfaces;
+    surfaces.count = twosided?2u:1u;
+    surfaces[0] = ir->allocNode<IR::INode>(IR::INode::ES_FRONT_SURFACE);
+    surfaces[0]->children = surfParent->children;
+    if (surfaces.count>1u) {
+        surfaces[1] = ir->allocNode<IR::INode>(IR::INode::ES_BACK_SURFACE);
+        surfaces[1]->children = surfParent->children;
+    }
+    surfParent->children = surfaces;
```

Comment:
I dont get this, exaplin?

Reply:
twosided is encoded in IR as presence of surfaces. Front surface is always present. Back surface is optional and if it is present, it means it's twosided

Reply:
twosided in mitsuba means it has the same BSDF on both sides... just saying.

Also if you allow this then you need a "fixed" front&back separate root node (two separate ASTs), not at literally every single node on the way down

Makes for a cleaner implementation, two separate instruction streams and stuff for front and back (but could be identical as in the case of mitsuba) then in GLSL or whatever you can do `instructionStream[gl_FrontFace ? frontOffset:backOffset]` instead of turning our IR into an abomination

Also get rid of dedicated root node type, I dont see a point of having it